httpbp/client_middlewares.go

```diff
@@ -200,16 +201,23 @@ func CircuitBreaker(config breakerbp.Config) ClientMiddleware {
 // Retries provides a retry middleware by ensuring certain HTTP responses are
 // wrapped in errors. Retries wraps the ClientErrorWrapper middleware, e.g. if
 // you are using Retries there is no need to also use ClientErrorWrapper.
-func Retries(limit int, retryOptions ...retry.Option) ClientMiddleware {
+func Retries(maxErrorReadAhead int, retryOptions ...retry.Option) ClientMiddleware {
 	if len(retryOptions) == 0 {
 		retryOptions = []retry.Option{retry.Attempts(1)}
 	}
 	return func(next http.RoundTripper) http.RoundTripper {
 		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			if req.GetBody != nil {
```

Comment:
that in most cases would result in 4xx errors anyways, I don't think we should try to get _too smart_.

maybe after the retry attempt, check if it's 4xx and the method is non-get and `GetBody` is nil, print a log message saying this is failed likely because you didn't set GetBody, but doing anything beyond the logging would be "too smart" in my books :)

also since `http.NewRequest`(-`WithContext`) auto populates `GetBody` with `*bytes.Buffer`/`*bytes.Reader`/`*strings.Reader` which should already covers like 95% of the cases how people give the request a body, I won't worry too much about that.

Reply:
a few things:
* we shouldn't assume that sending a request without a body is even safe, let alone that we'll be able to distinguish the error the server will return
* failing to have a GetBody can fail _even before_ you have to do a retry, if the request redirects
* there are tons of instances of [&http.Request{](https://sourcegraph.build.ue1.snooguts.net/search?patternType=regexp&q=context%3Areddit+lang%3Ago+%26http.Request&groupBy=repo) that are not calling NewRequest.  Honestly I rarely call it, because I don't like handling the error.

I think the right thing to do is to skip the middleware (with a log message indicating why) if we're not going to be able to function properly, since we can't do the _better_ thing and return an error immediately (since that would be a breaking change).

Reply:
I don't think there's anywhere in the RFC to say that a POST request must have a body. A server can be lenient on the method (want GET but don't complain if it's a POST instead) and the client can misuse POST instead of GET without a body. So being _too smart_ can break RFC compliant code (although definitely not ideal code).

But I guess rejecting requests with non-nil `Body` and nil `GetBody` is probably safe enough.

Reply:
I struggle to draw the line here between useful defaults and not trying to be too clever.

At least I can see a few cases where POST and nil-body makes sense:

* REST endpoint to create an entity but the server side creates it with default attributes, no body needed 
* Upload of an empty file

While the latter maybe should be an error I'd think it's not for the middleware to decide that. 

But maybe Kyle knows of instances where this just leads to more headaches.